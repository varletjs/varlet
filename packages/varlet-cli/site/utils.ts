import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
