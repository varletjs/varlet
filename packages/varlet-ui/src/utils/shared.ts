import { TimeData } from '../countdown/props'

export interface CacheInstance<T> {
  cache: T[]

  has(key: T): boolean

  add(key: T): void

  remove(key: T): void

  clear(): void
}

export const toNumber = (val: number | string | boolean | undefined | null): number => {
  if (val == null) return 0

  if (isString(val)) {
    val = parseFloat(val)
    val = Number.isNaN(val) ? 0 : val
    return val
  }

  if (isBool(val)) return Number(val)

  return val
}

export const isHTMLSupportImage = (val: string | undefined | null) => {
  if (val == null) {
    return false
  }

  return val.startsWith('data:image') || /\.(png|jpg|gif|jpeg|svg)$/.test(val)
}

export const isHTMLSupportVideo = (val: string | undefined | null) => {
  if (val == null) {
    return false
  }

  return val.startsWith('data:video') || /\.(mp4|webm|ogg)$/.test(val)
}

export const isString = (val: unknown): val is string => typeof val === 'string'

export const isBool = (val: unknown): val is boolean => typeof val === 'boolean'

export const isNumber = (val: unknown): val is number => typeof val === 'number'

export const isPlainObject = (val: unknown): val is Record<string, any> =>
  Object.prototype.toString.call(val) === '[object Object]'

export const isObject = (val: unknown): val is Record<string, any> => typeof val === 'object' && val !== null

export const isArray = (val: unknown): val is Array<any> => Array.isArray(val)

export const isURL = (val: string | undefined | null) => {
  if (!val) {
    return false
  }

  return /^(http)|(\.*\/)/.test(val)
}

export const isEmpty = (val: unknown) =>
  val === undefined || val === null || val === '' || (Array.isArray(val) && !val.length)

export const removeItem = (arr: Array<unknown>, item: unknown) => {
  if (arr.length) {
    const index: number = arr.indexOf(item)
    if (index > -1) {
      return arr.splice(index, 1)
    }
  }
}

export const throttle = (method: any, mustRunDelay = 200): (() => void) => {
  let timer: number
  let start = 0

  return function loop(this: unknown, ...args) {
    const now = Date.now()
    const elapsed = now - start

    if (!start) {
      start = now
    }

    if (timer) {
      window.clearTimeout(timer)
    }

    if (elapsed >= mustRunDelay) {
      method.apply(this, args)
      start = now
    } else {
      timer = window.setTimeout(() => {
        loop.apply(this, args)
      }, mustRunDelay - elapsed)
    }
  }
}

export const createCache = <T>(max: number): CacheInstance<T> => {
  const cache: T[] = []

  return {
    cache,

    has(key: T) {
      return this.cache.includes(key)
    },

    add(key: T) {
      if (this.has(key)) {
        return
      }

      this.cache.length === max && cache.shift()
      this.cache.push(key)
    },

    remove(key: T) {
      this.has(key) && removeItem(this.cache, key)
    },

    clear() {
      this.cache.length = 0
    },
  }
}

export const linear = (value: number): number => value

export const cubic = (value: number): number => value ** 3

export const easeInOutCubic = (value: number): number =>
  value < 0.5 ? cubic(value * 2) / 2 : 1 - cubic((1 - value) * 2) / 2

export function parseFormat(format: string, time: TimeData): string {
  const scannedTimes = Object.values(time)
  const scannedFormats = ['DD', 'HH', 'mm', 'ss']
  const padValues = [24, 60, 60, 1000]

  scannedFormats.forEach((scannedFormat, index) => {
    if (!format.includes(scannedFormat)) {
      scannedTimes[index + 1] += scannedTimes[index] * padValues[index]
    } else {
      format = format.replace(scannedFormat, String(scannedTimes[index]).padStart(2, '0'))
    }
  })

  if (format.includes('S')) {
    const ms = String(scannedTimes[scannedTimes.length - 1]).padStart(3, '0')

    if (format.includes('SSS')) {
      format = format.replace('SSS', ms)
    } else if (format.includes('SS')) {
      format = format.replace('SS', ms.slice(0, 2))
    } else {
      format = format.replace('S', ms.slice(0, 1))
    }
  }

  return format
}

export const dt = (value: unknown, defaultText: string | undefined) => (value == null ? defaultText : value)

export const inBrowser = () => typeof window !== 'undefined'

export const uniq = (arr: Array<any>) => [...new Set(arr)]

export function kebabCase(str: string): string {
  const ret = str.replace(/([A-Z])/g, ' $1').trim()
  return ret.split(' ').join('-').toLowerCase()
}
