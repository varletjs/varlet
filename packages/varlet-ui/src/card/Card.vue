<template>
  <div
    ref="card"
    :class="
      classes(
        n(),
        [isRow, n('--layout-row')],
        [outline, n('--outline')],
        [elevation, n(`$-elevation--${elevation}`), n('$-elevation--1')]
      )
    "
    :style="{
      zIndex: floated ? zIndex : undefined,
    }"
    @click="onClick"
    v-ripple="{ disabled: !ripple || floater }"
  >
    <div
      ref="cardFloater"
      :class="n('floater')"
      :style="{
        width: floaterWidth,
        height: floaterHeight,
        top: floaterTop,
        left: floaterLeft,
        overflow: floaterOverflow,
        position: floaterPosition,
        transition: floated
          ? `background-color ${floatingDuration}ms, color ${floatingDuration}ms, width ${floatingDuration}ms, height ${floatingDuration}ms, top ${floatingDuration}ms, left ${floatingDuration}ms`
          : undefined,
      }"
    >
      <slot name="image">
        <img
          :class="n('image')"
          :style="{
            objectFit: fit,
            height: toSizeUnit(imageHeight),
            width: toSizeUnit(imageWidth),
          }"
          :src="src"
          :alt="alt"
          v-if="src"
        />
      </slot>

      <div :class="n('container')">
        <slot name="title">
          <div :class="n('title')" v-if="title">{{ title }}</div>
        </slot>
        <slot name="subtitle">
          <div :class="n('subtitle')" v-if="subtitle">{{ subtitle }}</div>
        </slot>
        <slot name="description">
          <div :class="n('description')" v-if="description">{{ description }}</div>
        </slot>
        <div :class="n('footer')" v-if="$slots.extra">
          <slot name="extra" />
        </div>
        <div
          :class="n('floating-content')"
          :style="{
            height: contentHeight,
            opacity,
            transition: `opacity ${floatingDuration * 2}ms`,
          }"
          v-if="$slots['floating-content'] && !isRow"
        >
          <slot name="floating-content" />
        </div>
      </div>

      <div
        :class="classes(n('floating-buttons'), n('$--box'))"
        :style="{
          zIndex,
          opacity,
          transition: `opacity ${floatingDuration * 2}ms`,
        }"
        v-if="showFloatingButtons"
      >
        <slot name="close-button">
          <var-button var-card-cover round :class="classes(n('close-button'), n('$-elevation--6'))" @click.stop="close">
            <var-icon var-card-cover name="window-close" :class="n('close-button-icon')" />
          </var-button>
        </slot>
      </div>
    </div>

    <div
      :class="n('holder')"
      :style="{
        width: holderWidth,
        height: holderHeight,
      }"
    ></div>
  </div>
</template>

<script lang="ts">
import Ripple from '../ripple'
import VarIcon from '../icon'
import VarButton from '../button'
import { ref, defineComponent, watch, computed, nextTick } from 'vue'
import { props } from './props'
import { doubleRaf, toSizeUnit } from '../utils/elements'
import { call, createNamespace } from '../utils/components'
import { useZIndex } from '../context/zIndex'
import { useLock } from '../context/lock'
import type { Ref } from 'vue'

const { n, classes } = createNamespace('card')

const RIPPLE_DELAY = 500

export default defineComponent({
  name: 'VarCard',
  directives: { Ripple },
  components: {
    VarIcon,
    VarButton,
  },
  props,
  setup(props) {
    const card: Ref<null | HTMLElement> = ref(null)
    const cardFloater: Ref<null | HTMLElement> = ref(null)
    const holderWidth: Ref<string> = ref('auto')
    const holderHeight: Ref<string> = ref('auto')
    const floaterWidth: Ref<string> = ref('100%')
    const floaterHeight: Ref<string> = ref('100%')
    const floaterTop: Ref<string> = ref('auto')
    const floaterLeft: Ref<string> = ref('auto')
    const floaterPosition: Ref<string | undefined> = ref(undefined)
    const floaterOverflow: Ref<string> = ref('hidden')
    const contentHeight: Ref<string> = ref('0px')
    const opacity: Ref<string> = ref('0')
    const isRow = computed(() => props.layout === 'row')
    const showFloatingButtons: Ref<boolean> = ref(false)
    const floated: Ref<boolean> = ref(false)
    const { zIndex } = useZIndex(() => props.floating, 1)

    useLock(
      () => props.floating,
      () => !isRow.value
    )

    let dropdownFloaterTop = 'auto'
    let dropdownFloaterLeft = 'auto'
    let dropper: any = null
    const floater: any = ref(null)

    const floating = async () => {
      clearTimeout(floater.value)
      clearTimeout(dropper)

      floater.value = null

      floater.value = setTimeout(
        async () => {
          const { width, height, left, top } = card.value!.getBoundingClientRect()
          holderWidth.value = <string>toSizeUnit(width)
          holderHeight.value = <string>toSizeUnit(height)
          floaterWidth.value = holderWidth.value
          floaterHeight.value = holderHeight.value
          floaterTop.value = <string>toSizeUnit(top)
          floaterLeft.value = <string>toSizeUnit(left)
          floaterPosition.value = 'fixed'
          dropdownFloaterTop = floaterTop.value
          dropdownFloaterLeft = floaterLeft.value
          showFloatingButtons.value = true

          await doubleRaf()

          floaterTop.value = '0'
          floaterLeft.value = '0'
          floaterWidth.value = '100vw'
          floaterHeight.value = '100vh'
          contentHeight.value = 'auto'
          opacity.value = '1'
          floaterOverflow.value = 'auto'
          floated.value = true
        },
        props.ripple ? RIPPLE_DELAY : 0
      )
    }

    const dropdown = () => {
      clearTimeout(dropper)
      clearTimeout(floater.value)
      floater.value = null

      floaterWidth.value = holderWidth.value
      floaterHeight.value = holderHeight.value
      floaterTop.value = dropdownFloaterTop
      floaterLeft.value = dropdownFloaterLeft
      contentHeight.value = '0px'
      opacity.value = '0'
      showFloatingButtons.value = false

      dropper = setTimeout(() => {
        holderWidth.value = 'auto'
        holderHeight.value = 'auto'
        floaterWidth.value = '100%'
        floaterHeight.value = '100%'
        floaterTop.value = 'auto'
        floaterLeft.value = 'auto'
        dropdownFloaterTop = 'auto'
        dropdownFloaterLeft = 'auto'
        floaterOverflow.value = 'hidden'
        floaterPosition.value = undefined
        floated.value = false
      }, props.floatingDuration)
    }

    const close = () => {
      call(props['onUpdate:floating'], false)
    }

    watch(
      () => props.floating,
      (value) => {
        if (isRow.value) return

        nextTick(() => {
          value ? floating() : dropdown()
        })
      },
      { immediate: true }
    )

    return {
      n,
      classes,
      toSizeUnit,
      card,
      cardFloater,
      holderWidth,
      holderHeight,
      floater,
      floaterWidth,
      floaterHeight,
      floaterTop,
      floaterLeft,
      floaterPosition,
      floaterOverflow,
      contentHeight,
      opacity,
      zIndex,
      isRow,
      close,
      showFloatingButtons,
      floated,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../ripple/ripple';
@import './card';
</style>
