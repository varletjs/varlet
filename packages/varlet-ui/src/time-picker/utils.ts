import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
