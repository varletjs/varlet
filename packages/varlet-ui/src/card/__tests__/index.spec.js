import Card from '..'
import VarCard from '../Card'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay } from '../../utils/jest'

test('test card use', () => {
  const app = createApp({}).use(Card)
  expect(app.component(Card.name)).toBeTruthy()
})

describe('test card component events', () => {
  test('test card onClick with null callback', () => {
    const wrapper = mount(VarCard)
    wrapper.trigger('click')
    wrapper.unmount()
  })

  test('test card onClick', () => {
    const onClick = jest.fn()
    const wrapper = mount(VarCard, {
      props: {
        onClick,
      },
    })

    wrapper.trigger('click')
    expect(onClick).toHaveBeenCalledTimes(1)
    wrapper.unmount()
  })
})

describe('test card component props', () => {
  test('test card title', () => {
    const wrapper = mount(VarCard, {
      props: {
        title: 'This is title',
      },
    })

    expect(wrapper.find('.var-card__title').element.textContent).toBe('This is title')
    wrapper.unmount()
  })

  test('test card subtitle', () => {
    const wrapper = mount(VarCard, {
      props: {
        subtitle: 'This is subtitle',
      },
    })

    expect(wrapper.find('.var-card__subtitle').element.textContent).toBe('This is subtitle')
    wrapper.unmount()
  })

  test('test card description', () => {
    const wrapper = mount(VarCard, {
      props: {
        description: 'This is description',
      },
    })

    expect(wrapper.find('.var-card__description').element.textContent).toBe('This is description')
    wrapper.unmount()
  })

  test('test card elevation', () => {
    const wrapper = mount(VarCard, {
      props: {
        elevation: 3,
      },
    })

    expect(wrapper.find('.var-elevation--3').exists()).toBe(true)
    wrapper.unmount()
  })

  test('test card outline', () => {
    const wrapper = mount(VarCard, {
      props: {
        outline: true,
      },
    })

    expect(wrapper.find('.var-card--outline').exists()).toBe(true)
    wrapper.unmount()
  })

  test('test card src', () => {
    const wrapper = mount(VarCard, {
      props: {
        src: 'https://varlet.gitee.io/varlet-ui/cat.jpg',
      },
    })

    expect(wrapper.find('img').attributes('src')).toBe('https://varlet.gitee.io/varlet-ui/cat.jpg')
    wrapper.unmount()
  })

  test('test card fit', () => {
    ;['fill', 'contain', 'cover', 'none', 'scale-down'].forEach((fit) => {
      const wrapper = mount(VarCard, {
        props: {
          fit,
          src: 'https://varlet.gitee.io/varlet-ui/cat.jpg',
        },
      })

      expect(wrapper.find('img').attributes('style')).toContain('object-fit: ' + fit)
      wrapper.unmount()
    })
  })

  test('test card alt', () => {
    const wrapper = mount(VarCard, {
      props: {
        alt: 'This is alt',
        src: 'https://varlet.gitee.io/varlet-ui/cat.jpg',
      },
    })

    expect(wrapper.find('img').attributes('alt')).toBe('This is alt')
    wrapper.unmount()
  })

  test('test card height', async () => {
    const wrapper = mount(VarCard, {
      props: {
        imageHeight: 100,
        src: 'https://varlet.gitee.io/varlet-ui/cat.jpg',
      },
    })

    expect(wrapper.find('img').attributes('style')).toContain('height: 100px;')
    wrapper.unmount()
  })

  test('test card width', () => {
    const wrapper = mount(VarCard, {
      props: {
        imageWidth: 100,
        src: 'https://varlet.gitee.io/varlet-ui/cat.jpg',
      },
    })

    expect(wrapper.find('img').attributes('style')).toContain('width: 100px;')
    wrapper.unmount()
  })

  test('test card floating', async () => {
    const wrapper = mount(VarCard, {
      props: {
        floating: false,
      },
    })

    expect(wrapper.find('.var-card__floater').attributes('style')).toContain('width: 100%;')
    await wrapper.setProps({
      floating: true,
    })
    await delay(300)
    expect(wrapper.find('.var-card__floater').attributes('style')).toContain('width: 100vw;')
    wrapper.unmount()
  })
})
