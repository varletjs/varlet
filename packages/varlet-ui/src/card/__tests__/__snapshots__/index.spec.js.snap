// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test card example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\">
    <!--v-if-->
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <!--v-if-->
      <div class=\\"var-card__description\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</div>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>
<div class=\\"app-type\\">显示副标题</div>
<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\">
    <!--v-if-->
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <div class=\\"var-card__subtitle\\">我表情悠哉 跳个大概</div>
      <div class=\\"var-card__description\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</div>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>
<div class=\\"app-type\\">显示图片</div>
<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <div class=\\"var-card__subtitle\\">我表情悠哉 跳个大概</div>
      <div class=\\"var-card__description\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</div>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>
<div class=\\"app-type\\">使用插槽</div>
<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <div class=\\"var-card__subtitle\\">我表情悠哉 跳个大概</div>
      <div class=\\"var-card__description\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</div>
      <div class=\\"var-card__footer\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--primary var-elevation--2\\" style=\\"margin-right: 10px;\\">
          <!--v-if-->
          <div class=\\"var-button__content\\">添加按钮</div>
        </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--warning var-elevation--2\\">
          <!--v-if-->
          <div class=\\"var-button__content\\">添加按钮</div>
        </button></div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>
<div class=\\"app-type\\">横向显示</div>
<div class=\\"var-card var-card--layout-row\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <div class=\\"var-card__subtitle\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</div>
      <!--v-if-->
      <div class=\\"var-card__footer\\"><button class=\\"var-button var--box var-button--small var--inline-flex var-button--text-primary var-button--text\\">
          <!--v-if-->
          <div class=\\"var-button__content\\">添加按钮</div>
        </button></div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>
<div class=\\"app-type\\">水波效果</div>
<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\">
    <!--v-if-->
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <div class=\\"var-card__subtitle\\">我表情悠哉 跳个大概</div>
      <div class=\\"var-card__description\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</div>
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>
<div class=\\"app-type\\">全屏显示</div>
<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">本草纲目</div>
      <div class=\\"var-card__subtitle\\">我表情悠哉 跳个大概</div>
      <!--v-if-->
      <div class=\\"var-card__footer\\"><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-info var-button--text\\" style=\\"margin-right: 10px;\\">
          <!--v-if-->
          <div class=\\"var-button__content\\">添加按钮</div>
        </button><button class=\\"var-button var--box var-button--normal var--inline-flex var-button--text-info var-button--text\\">
          <!--v-if-->
          <div class=\\"var-button__content\\">添加按钮</div>
        </button></div>
      <div class=\\"var-card__content\\" style=\\"height: 0px; opacity: 0; transition: opacity 500ms;\\">
        <div class=\\"var-divider var--box\\">
          <!--v-if-->
        </div>
        <p style=\\"padding: 0px 14px;\\">如果华佗再世,崇洋都被医治,外邦来学汉字,激发我民族意识。马钱子、决明子、苍耳子，还有莲子；黄药子、苦豆子、川楝子，我要面子。用我的方式，改写一部历史。没什么别的事，跟着我念几个字。山药当归枸杞 GO，山药 当归 枸杞 GO，看我抓一把中药，服下一帖骄傲~</p>
      </div>
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>"
`;

exports[`test card floating 1`] = `
"<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">Title</div>
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <div class=\\"var-card__content\\" style=\\"height: 0px; opacity: 0; transition: opacity 500ms;\\">Content</div>
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>"
`;

exports[`test card floating 2`] = `
"<div class=\\"var-card var-card--layout-row\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms; top: 0px; left: 0px; border-radius: 0px;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">Title</div>
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>"
`;

exports[`test card layout 1`] = `
"<div class=\\"var-card var-card--layout-row\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">This is Card</div>
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>"
`;

exports[`test card layout 2`] = `
"<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--1\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">This is Card</div>
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>"
`;

exports[`test card props 1`] = `
"<div class=\\"var-card\\">
  <div class=\\"var-card__floater var-elevation--2\\" style=\\"width: 100%; height: 100%; overflow: hidden; position: static; transition: background-color 250ms, border-radius 250ms, width 250ms, height 250ms, top 250ms, left 250ms;\\"><img class=\\"var-card__image\\" style=\\"object-fit: cover; height: 200px; width: 200px;\\" src=\\"https://varlet-varletjs.vercel.app/cat.jpg\\" alt=\\"This is an image\\">
    <div class=\\"var-card__container\\">
      <div class=\\"var-card__title\\">This is Card</div>
      <div class=\\"var-card__subtitle\\">This is subtitle</div>
      <div class=\\"var-card__description\\">This is description</div>
      <div class=\\"var-card__footer\\">text</div>
      <!--v-if-->
    </div>
    <!--v-if-->
  </div>
  <div class=\\"var-card__holder\\" style=\\"width: auto; height: auto;\\"></div>
</div>"
`;
